File: Python/ml_methods/ridge_regression/demo_ridge_01_in_memory.py
# ==============================================================================
# demo_ridge_01_in_memory.py
# ==============================================================================
#
# Demo illustrating RidgeSolver usage on in-memory data.
# Mirrors cpp/ml_methods/ridge_regression/demo_mlm_ridge_01/demo_mlm_ridge_01.cpp.
#
# Demo content:
#   - Part A: Primal path (n = 500, p = 50, lambda = 1.0) — recover a dense
#             coefficient vector from a known linear model
#   - Part B: Dual path (n = 80, p = 300, lambda = 10.0) — high-dimensional
#             setting with a sparse ground truth; n < p routes through the
#             n x n dual system
#   - Part C: Lambda sweep (n = 200, p = 40) — coefficient error and residual
#             across lambda in {1e-4, ..., 1e3}, illustrating the shrinkage
#             bias-variance trade-off
#
# Notes:
#   - RidgeSolver.solve(X, y, lambda_val) is a static method solving
#     min_beta ||y - X beta||_2^2 + lambda * ||beta||_2^2. The L2 penalty
#     shrinks coefficients toward zero: larger lambda means more shrinkage
#     (more bias, less variance) and a larger residual on the training data.
#   - The C++ demo seeds std::mt19937; here we seed numpy's default_rng with
#     the same values. The RNG streams differ, so the printed numbers match
#     the C++ output qualitatively, not bitwise.
#
# ==============================================================================

import numpy as np
from trex_selector_neo.ml_methods import RidgeSolver

print("\n" + "=" * 70)
print("RidgeSolver Demos (in-memory)")
print("=" * 70 + "\n")


# ==============================================================================
# Part A: Basic Solve (n >= p, primal path)
# ==============================================================================

print("-" * 70)
print("Part A: RidgeSolver - Primal Path (n >= p)")
print("-" * 70 + "\n")

n, p = 500, 50
lam = 1.0
print(f"Matrix size : {n} x {p}")
print(f"Lambda      : {lam}\n")

rng = np.random.default_rng(42)
# The solver accesses the matrix through Eigen and therefore requires a
# Fortran-ordered (column-major) float64 array.
X = np.asfortranarray(rng.standard_normal((n, p)))

# Generate data from a known linear model: y = X * beta_true + noise.
beta_true = np.ones(p)
y = X @ beta_true + 0.1 * rng.standard_normal(n)

beta_hat = RidgeSolver.solve(X, y, lam)

# With plenty of samples (n >> p), low noise, and mild regularization the
# estimate lands very close to the ground truth.
coef_error = np.linalg.norm(beta_hat - beta_true)
residual = np.linalg.norm(y - X @ beta_hat)
print(f"||beta_hat - beta_true||_2 : {coef_error:.6e}")
print(f"Residual ||y - X*beta||_2  : {residual:.6e}")
assert coef_error < 0.5, "primal-path estimate should be close to beta_true"
print("\nPrimal-path estimate close to ground truth")


# ==============================================================================
# Part B: Dual Path (n < p)
# ==============================================================================

print("\n" + "-" * 70)
print("Part B: RidgeSolver - Dual Path (n < p)")
print("-" * 70 + "\n")

n, p = 80, 300
lam = 10.0
print(f"Matrix size : {n} x {p}")
print(f"Lambda      : {lam}\n")

rng = np.random.default_rng(7)
X = np.asfortranarray(rng.standard_normal((n, p)))

# Sparse ground truth: only the first 10 coefficients are nonzero. With n < p
# the least-squares problem is underdetermined; the ridge penalty makes it
# well-posed, and the solver switches to the cheaper n x n dual formulation.
beta_true = np.zeros(p)
beta_true[:10] = 1.0
y = X @ beta_true + 0.1 * rng.standard_normal(n)

beta_hat = RidgeSolver.solve(X, y, lam)

coef_error = np.linalg.norm(beta_hat - beta_true)
residual = np.linalg.norm(y - X @ beta_hat)
print(f"||beta_hat - beta_true||_2 : {coef_error:.6e}")
print(f"Residual ||y - X*beta||_2  : {residual:.6e}")


# ==============================================================================
# Part C: Lambda Sweep
# ==============================================================================

print("\n" + "-" * 70)
print("Part C: RidgeSolver - Lambda Sweep")
print("-" * 70 + "\n")

n, p = 200, 40
print(f"Matrix size : {n} x {p}\n")

rng = np.random.default_rng(99)
X = np.asfortranarray(rng.standard_normal((n, p)))

beta_true = np.ones(p)
y = X @ beta_true + 0.5 * rng.standard_normal(n)

# Sweep over a range of lambda values. As lambda grows, the penalty dominates
# the fit: coefficients shrink toward zero (||beta_hat||_2 decreases
# monotonically) and the training residual grows.
lambdas = [1e-4, 1e-2, 0.1, 1.0, 10.0, 100.0, 1000.0]

print(f"{'lambda':>12}  {'||beta_hat - true||_2':>22}  {'||residual||_2':>16}  {'||beta_hat||_2':>16}")
print("-" * 72)

beta_norms = []
residuals = []
for lam in lambdas:
    beta_hat = RidgeSolver.solve(X, y, lam)
    coef_error = np.linalg.norm(beta_hat - beta_true)
    residual = np.linalg.norm(y - X @ beta_hat)
    beta_norms.append(np.linalg.norm(beta_hat))
    residuals.append(residual)
    print(f"{lam:>12.4e}  {coef_error:>22.6e}  {residual:>16.6e}  {beta_norms[-1]:>16.6e}")

# Didactic shrinkage checks.
assert np.all(np.diff(beta_norms) < 0), "||beta_hat||_2 must shrink as lambda grows"
assert np.all(np.diff(residuals) > 0), "training residual must grow as lambda grows"
print("\nShrinkage check passed: ||beta_hat||_2 decreases and the residual")
print("increases monotonically in lambda.")

print("\n" + "=" * 70)
print("demo_ridge_01_in_memory.py complete.")

File: R/ml_methods/ridge_regression/demo_ridge_01_in_memory.R
# ==============================================================================
# demo_ridge_01_in_memory.R
# ==============================================================================
#
# Demo illustrating RidgeSolver on in-memory data.
#
# Demo content:
#   - Part 1: Primal path (n >= p) — recovering a dense linear model
#   - Part 2: Dual path (n < p) — high-dimensional, sparse ground truth
#   - Part 3: Lambda sweep — coefficient shrinkage as the penalty grows
#
# Mirrors: cpp/ml_methods/ridge_regression/demo_mlm_ridge_01/demo_mlm_ridge_01.cpp
#
# Note: same n / p / lambda values and seeds as the C++ demo; R's
# Mersenne-Twister stream differs from std::mt19937, so results match
# qualitatively, not bitwise.
#
# ==============================================================================

library(TRexSelectorNeo)

# ==============================================================================

cat("\n", strrep("=", 70), "\n", sep = "")
cat("RidgeSolver Demos (in-memory)\n")
cat(strrep("=", 70), "\n\n")

# Ridge regression solves the penalized least-squares problem
#   beta_hat = argmin ||y - X beta||_2^2 + lambda ||beta||_2^2 .
# The solver picks its computational path automatically:
#   - primal (p x p system) when n >= p,
#   - dual   (n x n system) when n <  p — same solution, cheaper factorization.

solver <- RidgeSolver$new()

l2_norm <- function(v) sqrt(sum(v^2))

# ==============================================================================
# Part 1: Basic Solve (n >= p, primal path)
# ==============================================================================
# Dense ground truth (all coefficients = 1), low noise: with mild
# regularization the ridge estimate should sit close to beta_true.
# ==============================================================================

cat(strrep("-", 70), "\n")
cat("Part 1: Primal Path (n >= p)\n")
cat(strrep("-", 70), "\n\n")

set.seed(42)
n1      <- 500L
p1      <- 50L
lambda1 <- 1.0

cat(sprintf("Matrix size : %d x %d\n", n1, p1))
cat(sprintf("Lambda      : %g\n\n", lambda1))

X1         <- matrix(rnorm(n1 * p1), nrow = n1, ncol = p1)
beta_true1 <- rep(1.0, p1)
y1         <- as.numeric(X1 %*% beta_true1) + 0.1 * rnorm(n1)

beta_hat1 <- solver$solve(X1, y1, lambda1)

coef_error1 <- l2_norm(beta_hat1 - beta_true1)
residual1   <- l2_norm(y1 - X1 %*% beta_hat1)
cat("||beta_hat - beta_true||_2 :", format(coef_error1, scientific = TRUE), "\n")
cat("Residual ||y - X*beta||_2  :", format(residual1, scientific = TRUE), "\n")

# Didactic check: with sd(noise) = 0.1 the estimate must be close to truth.
stopifnot(coef_error1 < 0.5)
cat("\nCheck passed: ridge estimate close to the true coefficients.\n")

# ==============================================================================
# Part 2: Dual Path (n < p)
# ==============================================================================
# High-dimensional regime: only the first 10 of 300 coefficients are nonzero.
# Ridge does not select variables — it shrinks everything — so the estimate
# is biased, but the dual formulation keeps the problem well posed for n < p.
# ==============================================================================

cat("\n", strrep("-", 70), "\n")
cat("Part 2: Dual Path (n < p)\n")
cat(strrep("-", 70), "\n\n")

set.seed(7)
n2      <- 80L
p2      <- 300L
lambda2 <- 10.0

cat(sprintf("Matrix size : %d x %d\n", n2, p2))
cat(sprintf("Lambda      : %g\n\n", lambda2))

X2 <- matrix(rnorm(n2 * p2), nrow = n2, ncol = p2)

# Sparse ground truth: only the first 10 coefficients are nonzero.
beta_true2 <- c(rep(1.0, 10L), rep(0.0, p2 - 10L))
y2         <- as.numeric(X2 %*% beta_true2) + 0.1 * rnorm(n2)

beta_hat2 <- solver$solve(X2, y2, lambda2)

coef_error2 <- l2_norm(beta_hat2 - beta_true2)
residual2   <- l2_norm(y2 - X2 %*% beta_hat2)
cat("||beta_hat - beta_true||_2 :", format(coef_error2, scientific = TRUE), "\n")
cat("Residual ||y - X*beta||_2  :", format(residual2, scientific = TRUE), "\n")

# ==============================================================================
# Part 3: Lambda Sweep
# ==============================================================================
# As lambda grows, ||beta_hat||_2 shrinks monotonically toward 0 while the
# training residual grows — the classic bias-variance trade-off knob.
# ==============================================================================

cat("\n", strrep("-", 70), "\n")
cat("Part 3: Lambda Sweep\n")
cat(strrep("-", 70), "\n\n")

set.seed(99)
n3 <- 200L
p3 <- 40L

cat(sprintf("Matrix size : %d x %d\n\n", n3, p3))

X3         <- matrix(rnorm(n3 * p3), nrow = n3, ncol = p3)
beta_true3 <- rep(1.0, p3)
y3         <- as.numeric(X3 %*% beta_true3) + 0.5 * rnorm(n3)

lambdas <- c(1e-4, 1e-2, 0.1, 1.0, 10.0, 100.0, 1000.0)

cat(sprintf("%12s  %20s  %16s  %16s\n",
            "lambda", "||beta_hat - true||_2", "||beta_hat||_2", "||residual||_2"))
cat(strrep("-", 70), "\n")

beta_norms <- numeric(length(lambdas))
residuals  <- numeric(length(lambdas))
for (i in seq_along(lambdas)) {
  beta_hat      <- solver$solve(X3, y3, lambdas[i])
  beta_norms[i] <- l2_norm(beta_hat)
  residuals[i]  <- l2_norm(y3 - X3 %*% beta_hat)
  cat(sprintf("%12.4e  %20.6e  %16.6e  %16.6e\n",
              lambdas[i], l2_norm(beta_hat - beta_true3),
              beta_norms[i], residuals[i]))
}

# Didactic checks: for ridge, ||beta_hat(lambda)||_2 is non-increasing and
# the training residual is non-decreasing in lambda.
stopifnot(all(diff(beta_norms) <= 1e-10))
stopifnot(all(diff(residuals) >= -1e-10))
cat("\nCheck passed: coefficients shrink and residual grows with lambda.\n")

cat("\n", strrep("=", 70), "\n")
cat("demo_ridge_01_in_memory.R complete.\n")
